\begin{figure}[h!]
\includegraphics[width=0.75\textwidth]{ncbi/out/ncbi.png}
\centering
\caption{\textbf{Cumulative number of different eukaryotic genomes annotated by NCBI.}}
\label{sfig:ncbi}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{clusters/out/clusters.png}
\centering
\caption{\textbf{Statistics of orthologous groups.}
\textbf{(A)} Each species is equally represented in orthologous groups (OGs). \textbf{(B)} Nearly all proteins are associated with a single orthologous group. \textbf{(C)} A plurality of orthologous groups contain all species. \textbf{(D)} The number of orthologous groups associated with a species is strongly correlated with the number of unique annotated proteins, which suggests the annotation pipeline generally identifies conserved genes.}
\label{sfig:clusters}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{paralogs/out/paralogs.png}
\centering
\caption{\textbf{Addition of paralogs to orthologous groups.}
\textbf{(A)} Most orthologous groups (OGs) have no in-paralogs. \textbf{(B, D)} Of the groups with paralogs, most have fewer than five. \textbf{(C)} The in-paralogs are generally only a small fraction of the sequences in an orthologous group.}
\label{sfig:paralogs}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{insertion_training/out/insertion_training.png}
\centering
\caption{\textbf{Insertion phylo-HMM data and training details.}
\textbf{(A)} Most columns in the training data were labeled as state 1A or 1B. \textbf{(B)} The model loss stabilized by the final training iteration. \textbf{(C-F)} The values of parameters in the phylogenetic process, the jump process, the pattern stickiness model, and the transition matrix, respectively, at each training iteration. The transition matrix plots are the transition rates to the state indicated on the vertical axis and given in log scale. Self transitions are excluded.}
\label{sfig:insertion_training}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{insertion_trim/out/insertion_trim.png}
\centering
\caption{\textbf{Insertion phylo-HMM trimming details.}
\textbf{(A)} Most alignments were not trimmed. Of the alignments with trims, most were trimmed only at the level of sequences. \textbf{(B)} Most trimmed regions were inferred primarily as state 2. \textbf{(C)} Most alignments with sequence trims have fewer than 10 segments removed. \textbf{(D)} Most alignments with region trims have fewer than five regions removed. \textbf{(E, G)} The number of non-gap symbols in sequence trims can vary considerably, but for nearly all sequence trims each non-gap symbol in the removed segment is aligned to fewer than five non-gap symbols on average. Only the lower 95\% of the distribution of the number of non-gap symbols in the sequence trims is shown. \textbf{(F, H)} The length of region trims can also vary considerably, but generally each region trims accounts for fewer than 10\% of the columns in the original alignment.}
\label{sfig:insertion_trim}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{missing_training/out/missing_training.png}
\centering
\caption{\textbf{Missing phylo-HMM data and training details.}
\textbf{(A)} Most columns in the training data were labeled as state 1, which is referred to as the ``not missing'' state in the main text. \textbf{(B)} The model loss stabilized by the final training iteration. \textbf{(C-F)} The values of parameters in the phylogenetic process, the jump process, and the transition matrix, respectively, at each training iteration. The transition matrix plots are the transition rates to the state indicated on the vertical axis and given in log scale. Self transitions are excluded.}
\label{sfig:missing_training}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{missing_trim/out/missing_trim.png}
\centering
\caption{\textbf{Missing phylo-HMM trimming details.}
\textbf{(A)} Most alignments have no sequences with ``missing'' segments. \textbf{(B)} Of the alignments with sequences trimmed of missing segments, a majority have only one trimmed sequence. \textbf{(C-D)} The length of missing segments can vary considerably, both in terms of the number of positions as well as its ratio to the number of columns in the alignment.}
\label{sfig:missing_trim}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{trees/out/trees2_LG.png}
\centering
\caption{\textbf{Phylogenetic trees fit to meta-alignments yielded by different sampling strategies under LG model.}}
\label{sfig:trees_LG}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{trees/out/trees2_GTR.png}
\centering
\caption{\textbf{Phylogenetic trees fit by different sampling strategies under GTR model.}}
\label{sfig:trees_GTR}
\end{figure}

\begin{table}[h!]
\centering
\caption{\textbf{Genome annotations.}}
\label{stable:genomes}
\begin{tabular}{|l|l|l|l|l|}
\hline
\textbf{Species}                       & \textbf{Species ID} & \textbf{Taxon ID} & \textbf{Version} & \textbf{Source} \\ \hline
\textit{Drosophila ananassae}          & dana                & 7217              & 102              & NCBI            \\ \hline
\textit{Drosophila biarmipes}          & dbia                & 125945            & 102              & NCBI            \\ \hline
\textit{Drosophila bipectinata}        & dbip                & 42026             & 102              & NCBI            \\ \hline
\textit{Drosophila elegans}            & dele                & 30023             & 102              & NCBI            \\ \hline
\textit{Drosophila erecta}             & dere                & 7220              & 101              & NCBI            \\ \hline
\textit{Drosophila eugracilis}         & deug                & 29029             & 102              & NCBI            \\ \hline
\textit{Drosophila ficusphila}         & dfic                & 30025             & 102              & NCBI            \\ \hline
\textit{Drosophila grimshawi}          & dgri                & 7222              & 103              & NCBI            \\ \hline
\textit{Drosophila guanche}            & dgua                & 7266              & 100              & NCBI            \\ \hline
\textit{Drosophila hydei}              & dhyd                & 7224              & 101              & NCBI            \\ \hline
\textit{Drosophila innubila}           & dinn                & 198719            & 100              & NCBI            \\ \hline
\textit{Drosophila kikkawai}           & dkik                & 30033             & 102              & NCBI            \\ \hline
\textit{Drosophila mauritiana}         & dmau                & 7226              & 100              & NCBI            \\ \hline
\textit{Drosophila melanogaster}       & dmel                & 7227              & FB2022\_02       & FlyBase         \\ \hline
\textit{Drosophila mojavensis}         & dmoj                & 7230              & 102              & NCBI            \\ \hline
\textit{Drosophila navojoa}            & dnav                & 7232              & 101              & NCBI            \\ \hline
\textit{Drosophila novamexicana}       & dnov                & 47314             & 100              & NCBI            \\ \hline
\textit{Drosophila obscura}            & dobs                & 7282              & 101              & NCBI            \\ \hline
\textit{Drosophila persimilis}         & dper                & 7234              & 101              & NCBI            \\ \hline
\textit{Drosophila pseudoobscura}      & dpse                & 7237              & 104              & NCBI            \\ \hline
\textit{Drosophila rhopaloa}           & drho                & 1041015           & 102              & NCBI            \\ \hline
\textit{Drosophila santomea}           & dsan                & 129105            & 101              & NCBI            \\ \hline
\textit{Drosophila sechellia}          & dsec                & 7238              & 101              & NCBI            \\ \hline
\textit{Drosophila serrata}            & dser                & 7274              & 100              & NCBI            \\ \hline
\textit{Drosophila simulans}           & dsim                & 7240              & 103              & NCBI            \\ \hline
\textit{Drosophila subobscura}         & dsob                & 7241              & 100              & NCBI            \\ \hline
\textit{Drosophila subpulchrella}      & dspu                & 1486046           & 100              & NCBI            \\ \hline
\textit{Drosophila suzukii}            & dsuz                & 28584             & 102              & NCBI            \\ \hline
\textit{Drosophila takahashii}         & dtak                & 29030             & 102              & NCBI            \\ \hline
\textit{Drosophila teissieri}          & dtei                & 7243              & 100              & NCBI            \\ \hline
\textit{Drosophila virilis}            & dvir                & 7244              & 103              & NCBI            \\ \hline
\textit{Drosophila willistoni}         & dwil                & 7260              & 102              & NCBI            \\ \hline
\textit{Drosophila yakuba}             & dyak                & 7245              & 102              & NCBI            \\ \hline
\textit{Scaptodrosophila lebanonensis} & sleb                & 7225              & 100              & NCBI            \\ \hline
\end{tabular}
\end{table}

\begin{table}[h!]
\centering
\caption{\textbf{Phylogenetic diversity criteria.}}
\label{stable:diversity_criteria}
\begin{tabular}{|l|l|}
\hline
\textbf{Species IDs}   & \textbf{Minimum number}   \\ \hline
dinn,   dgri, dhyd     & 2                         \\ \hline
dnov,   dvir           & 1                         \\ \hline
dmoj,   dnav           & 1                         \\ \hline
dper,   dpse           & 1                         \\ \hline
dgua,   dsob           & 1                         \\ \hline
dana,   dbip           & 1                         \\ \hline
dkik,   dser           & 1                         \\ \hline
dele,   drho           & 1                         \\ \hline
dtak,   dbia           & 1                         \\ \hline
dspu,   dsuz           & 1                         \\ \hline
dere,   dtei           & 1                         \\ \hline
dsan,   dyak           & 1                         \\ \hline
dmel                   & 1                         \\ \hline
dmau,   dsec, dsim     & 1                         \\ \hline
\end{tabular}
\end{table}
